### ISSUE TITLE
CPU memory slowly filling, during inference, cannot seem to find the reason

### ISSUE THREAD

**Issue by StevenPuttemans** [tensorflow/tensorflow#23348]
**System information**
- We have made a specific adaptation of a inference script, that reads in tensorflow records, pushes the data through the network and then stores the result.
- Linux Ubuntu 18.04
- TensorFlow installed from (source or binary): default Ubuntu repositories
- TensorFlow version (use command below): 1.3
- Python version: 2.7
- CUDA/cuDNN version: Cuda 8.0 CuDNN 6.0
- GPU model and memory: NVIDIA P5000 - 16GB

**Describe the current behavior**

The script runs slow but fine. However, CPU memory of the system is slowly clogging up and system will fail.

**Describe the expected behavior**

We expected the code to run through all data parts without issues.

**Code to reproduce the issue**

```
"""Applies model to all examples in a train record.

"""

import os
import numpy as np
import tensorflow as tf

import constants
import model
import util

_RESOLUTIONS = ['5cm', '9cm', '19cm']
_INPUT_FEATURE = 'input_sdf'
_TARGET_FEATURE = 'target_df'
_TARGET_SEM_FEATURE = 'target_sem'
_TRAIN_FEATURE = 'samples'
_TRAIN_SEM_FEATURE = 'samples_sem'

flags = tf.flags
FLAGS = flags.FLAGS
flags.DEFINE_string('base_dir', '',
                    'Root directory. Expects a directory containing the model.')
flags.DEFINE_string('model_checkpoint', '',
                    'Model checkpoint to use (empty for latest).')
flags.DEFINE_string('data_filepattern', '/tmp/data/train_*.tfrecords',
                    'Training data file pattern.')
flags.DEFINE_string('output_folder', FLAGS.base_dir + '/data/vox5-9-19-dim3D_edited/',
                    'Folder in which to put the new tfrecords.')

# Parameters for applying model.
flags.DEFINE_integer('height_input', 64, 'Input block y dim.')
flags.DEFINE_integer('hierarchy_level', 1, 'Hierachy level (1: finest level).')
flags.DEFINE_bool('is_base_level', True, 'If base level of hierarchy.')
flags.DEFINE_integer('num_quant_levels', 256, 'Number of quantization bins.')
flags.DEFINE_integer('num_total_hierarchy_levels', 1,
                     'Number of total hierarchy levels.')
flags.DEFINE_integer('stored_dim_block', 64,
                     'Stored data block x/z dim, high-resolution.')
flags.DEFINE_integer('stored_height_block', 64,
                     'Stored data block y dim, high-resolution.')
flags.DEFINE_integer('p_norm', 1, 'P-norm loss (0 to disable).')
flags.DEFINE_bool('predict_semantics', True,
                  'Also predict semantic labels per-voxel.')
flags.DEFINE_float('temperature', 100.0, 'Softmax temperature for sampling.')


def read_record(filename):
    for serialized_example in tf.python_io.tf_record_iterator(filename):
	samples_lo = np.zeros([FLAGS.stored_dim_block // 2, FLAGS.stored_height_block // 2,
                               FLAGS.stored_dim_block // 2, 2])
        samples_sem_lo = np.zeros([FLAGS.stored_dim_block // 2, FLAGS.stored_height_block // 2,
                                   FLAGS.stored_dim_block // 2], dtype=np.uint8)
        if not FLAGS.is_base_level:
            key_samples = 'samples_' + _RESOLUTIONS[FLAGS.hierarchy_level]
            key_samples_sem = 'sem_samples_' + _RESOLUTIONS[FLAGS.hierarchy_level]
            spec = {
                'data':
                    tf.FixedLenFeature((), tf.string),
                key_samples:
                    tf.FixedLenFeature((FLAGS.stored_dim_block // 2, FLAGS.stored_height_block // 2,
                                        FLAGS.stored_dim_block // 2, 1), tf.float32),
                key_samples_sem:
                    tf.FixedLenFeature((), tf.string)
            }
            example = tf.parse_single_example(serialized_example, spec)
            samples_lo = example[key_samples]
            samples_sem_lo = example[key_samples_sem]
            samples_sem_lo = tf.decode_raw(samples_sem_lo, tf.uint8)
            samples_sem_lo = tf.reshape(samples_sem_lo, [
                FLAGS.stored_dim_block // 2, FLAGS.stored_height_block // 2, FLAGS.stored_dim_block // 2])

            serialized_example = example['data']
	
        # Parse sequence example.
        key_input = _RESOLUTIONS[FLAGS.hierarchy_level - 1] + '_' + _INPUT_FEATURE
        key_target = _RESOLUTIONS[FLAGS.hierarchy_level - 1] + '_' + _TARGET_FEATURE
        key_target_sem = _RESOLUTIONS[FLAGS.hierarchy_level - 1] + '_' + _TARGET_SEM_FEATURE
	
        sequence_features_spec = {
            key_input:
                tf.FixedLenFeature(
                    (FLAGS.stored_dim_block, FLAGS.stored_height_block, FLAGS.stored_dim_block, 1),
                    tf.float32),
            key_target:
                tf.FixedLenFeature(
                    (FLAGS.stored_dim_block, FLAGS.stored_height_block, FLAGS.stored_dim_block, 1),
                    tf.float32),
            key_target_sem:
                tf.FixedLenFeature((), tf.string)
        }
	print "did it"
        example = tf.parse_single_example(serialized_example, sequence_features_spec)
	print "did it"
        input_sdf_tensor = example[key_input]
	
        input_sdf = input_sdf_tensor.eval()
        input_sdf = np.squeeze(input_sdf, axis = 3)
        input_sdf = util.preprocess_sdf(input_sdf, constants.TRUNCATION)

        yield input_sdf, samples_lo, samples_sem_lo, serialized_example


def predict_from_model(logit_groups_geometry, logit_groups_semantics,
                       temperature):
    """Reconstruct predicted geometry and semantics from model output."""
    predictions_geometry_list = []
    for logit_group in logit_groups_geometry:
        if FLAGS.p_norm > 0:
            predictions_geometry_list.append(logit_group[:, :, :, :, 0])
        else:
            logit_group_shape = logit_group.shape_as_list()
            logit_group = tf.reshape(logit_group, [-1, logit_group_shape[-1]])
            samples = tf.multinomial(temperature * logit_group, 1)
            predictions_geometry_list.append(
                tf.reshape(samples, logit_group_shape[:-1]))
    predictions_semantics_list = []
    if FLAGS.predict_semantics:
        for logit_group in logit_groups_semantics:
            predictions_semantics_list.append(tf.argmax(logit_group, 4))
    else:
        predictions_semantics_list = [
                                         tf.zeros(shape=predictions_geometry_list[0].shape, dtype=tf.uint8)
                                     ] * len(predictions_geometry_list)
    return predictions_geometry_list, predictions_semantics_list


def create_dfs_from_output(output_df):
    """Rescales model output to distance fields (in voxel units)."""
    if FLAGS.p_norm > 0:
        output_df = constants.TRUNCATION * (output_df[0, :, :, :, 0] + 1)
    else:
        output_df = (output_df[0, :, :, :, 0] + 1) * 0.5 * (
            FLAGS.num_quant_levels - 1)
        output_df = util.dequantize(output_df, FLAGS.num_quant_levels,
                                    constants.TRUNCATION)
    return output_df


def create_model(scene_dim_x, scene_dim_y, scene_dim_z):
    """Init model graph for scene."""
    input_placeholder = tf.placeholder(
        tf.float32,
        shape=[1, scene_dim_z, scene_dim_y, scene_dim_x, 2],
        name='pl_scan')
    target_placeholder = tf.placeholder(
        tf.float32,
        shape=[1, scene_dim_z, scene_dim_y, scene_dim_x, 2],
        name='pl_target')
    target_lo_placeholder = tf.placeholder(
        tf.float32,
        shape=[1, scene_dim_z // 2, scene_dim_y // 2, scene_dim_x // 2, 2],
        name='pl_target_lo')
    target_sem_placeholder = tf.placeholder(
        tf.uint8,
        shape=[1, scene_dim_z, scene_dim_y, scene_dim_x],
        name='pl_target_sem')
    target_sem_lo_placeholder = tf.placeholder(
        tf.uint8,
        shape=[1, scene_dim_z // 2, scene_dim_y // 2, scene_dim_x // 2],
        name='pl_target_sem_lo')
    # No previous level input if at base level.
    if FLAGS.is_base_level:
        target_scan_low_resolution = None
        target_semantics_low_resolution = None
    else:
        target_scan_low_resolution = target_lo_placeholder
        target_semantics_low_resolution = target_sem_lo_placeholder
    logits = model.model(
        input_scan=input_placeholder,
        target_scan_low_resolution=target_scan_low_resolution,
        target_scan=target_placeholder,
        target_semantics_low_resolution=target_semantics_low_resolution,
        target_semantics=target_sem_placeholder,
        num_quant_levels=FLAGS.num_quant_levels,
        predict_semantics=FLAGS.predict_semantics,
        use_p_norm=FLAGS.p_norm > 0)
    return (input_placeholder, target_placeholder, target_lo_placeholder,
            target_sem_placeholder, target_sem_lo_placeholder, logits)


def main(_):
    model_path = FLAGS.base_dir
    output_folder = FLAGS.output_folder
    if not os.path.exists(output_folder):
        os.makedirs(output_folder)

    if FLAGS.model_checkpoint:
        checkpoint_path = os.path.join(model_path, FLAGS.model_checkpoint)
        print "Used model: " + str(checkpoint_path)
    else:
        checkpoint_path = tf.train.latest_checkpoint(model_path)
        print "Used model: " + str(checkpoint_path)	

    # Loop over all tfrecords in training data
    data_filepattern = FLAGS.data_filepattern
    if not isinstance(data_filepattern, list):
        data_filepattern = [data_filepattern]
    # Get filenames matching filespec.
    tf.logging.info('data_filepattern: %s', data_filepattern)
    filenames = []
    for p in data_filepattern:
        filenames.extend(tf.gfile.Glob(p))
    tf.logging.info('filenames: %s', filenames)

    for filename in filenames:
        tf.reset_default_graph()

        # Init model.
        (input_placeholder, target_placeholder, target_lo_placeholder,
         target_sem_placeholder, target_sem_lo_placeholder, logits) = create_model(
            FLAGS.stored_dim_block, FLAGS.stored_height_block, FLAGS.stored_dim_block)
        logit_groups_geometry = logits['logits_geometry']
        logit_groups_semantics = logits['logits_semantics']
        feature_groups = logits['features']

        predictions_geometry_list, predictions_semantics_list = predict_from_model(
            logit_groups_geometry, logit_groups_semantics, FLAGS.temperature)


        config = tf.ConfigProto(device_count={'GPU': 0})
        init_op = tf.group(tf.global_variables_initializer(),
                           tf.local_variables_initializer())

        with tf.Session(config=config) as session:
            session.run(init_op)
	    
            assign_fn = tf.contrib.framework.assign_from_checkpoint_fn(
                checkpoint_path, tf.contrib.framework.get_variables_to_restore())
            assign_fn(session)
   
            filename_out = os.path.join(output_folder, os.path.basename(filename))
            print(filename_out)
            writer = tf.python_io.TFRecordWriter(filename_out)

            for input_sdf, samples_lo, samples_sem_lo, serialized_example in read_record(filename):             

		# Make batch size 1 to input to model.
                input_sdf = input_sdf[np.newaxis, :, :, :, :]
                samples_lo = samples_lo[np.newaxis, :, :, :, :]
                samples_sem_lo = samples_sem_lo[np.newaxis, :, :, :]
                output_df = np.ones(shape=input_sdf.shape)
                output_df[:, :, :, :, 0] *= constants.TRUNCATION  # Fill with truncation, known values.
                output_sem = np.zeros(
                    shape=[1, FLAGS.stored_dim_block, FLAGS.stored_height_block, FLAGS.stored_dim_block],
                    dtype=np.uint8)

                # First get features.
                feed_dict = {
                    input_placeholder: input_sdf,
                    target_lo_placeholder: samples_lo,
                    target_placeholder: output_df,
                    target_sem_lo_placeholder: samples_sem_lo,
                    target_sem_placeholder: output_sem
                }
		
                # Cache these features.
                feature_groups_ = session.run(feature_groups, feed_dict)

                for n in range(8):
                    tf.logging.info('Predicting group [%d/%d]', n + 1, 8)
                    # Predict
                    feed_dict[feature_groups[n]] = feature_groups_[n]
                    predictions = session.run(
                        {
                            'prediction_geometry': predictions_geometry_list[n],
                            'prediction_semantics': predictions_semantics_list[n]
                        },
                        feed_dict=feed_dict)
                    prediction_geometry = predictions['prediction_geometry']
                    prediction_semantics = predictions['prediction_semantics']
                    # Put into [-1,1] for next group.
                    if FLAGS.p_norm == 0:
                        prediction_geometry = prediction_geometry.astype(np.float32) / (
                            (FLAGS.num_quant_levels - 1) / 2.0) - 1.0

                    util.assign_voxel_group(output_df, prediction_geometry,
                                            n + 1)
                    if FLAGS.predict_semantics:
                        util.assign_voxel_group(output_sem,
                                                prediction_semantics, n + 1)

                # Final outputs.
                output_sem = output_sem[0]
                output_df = create_dfs_from_output(output_df)  # Fill with truncation, known values.

                # Export predictions
                key_samples = 'samples_' + _RESOLUTIONS[FLAGS.hierarchy_level - 1]
                key_samples_sem = 'sem_samples_' + _RESOLUTIONS[FLAGS.hierarchy_level - 1]
                out_feature = {
                    'data': util.bytes_feature(serialized_example),
                    key_samples: util.float_feature(output_df.flatten().tolist()),
                    key_samples_sem: util.bytes_feature(output_sem.flatten().tobytes())
                }
                example = tf.train.Example(features=tf.train.Features(feature=out_feature))
                writer.write(example.SerializeToString())
        session.close()

if __name__ == '__main__':
    tf.app.run(main)
```

There are no errors reported. Any help welcome.

**Comment by ragulpr** [comment]
Could you report something like the output of [memory_profiler](https://pypi.org/project/memory_profiler/)? 

Random stab at the problem:
I would expect system memory to explode once you empty gpu to cpu. Another is if  `feature_groups = logits['features']` is some large amount of data then `feed_dict[feature_groups[n]] = feature_groups_[n]` you'll slowly append things to the `feed_dict`

**Comment by harshini-gadige** [comment]
@reedwm  PTAL

**Comment by StevenPuttemans** [comment]
The issues is still valid I guess. On our end we solved it by using the Dataset interface provided in Tensorflow, which served our needs more. It seems the  original code simply did not take into account that used datasets can be lots bigger than what they used.

**Comment by ragulpr** [comment]
It would be helpful if you have a reproducible example that does not rely on your own dataset and reports on when/where it fails/your memory usage increases

**Comment by tensorflowbutler** [comment]
Nagging Assignee @azaks2: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mohantym** [comment]
Hi @StevenPuttemans! We are  checking to see if you are still looking for assistance in this issue.
Could you please try on latest stable version i.e  TF 2.6  and let us know if this is still an issue.Thanks!
